File: app/src/main/java/omniex/nl/omniex/ui/app/order/shipping/OrderShippingPresenter.java
package omniex.nl.omniex.ui.app.order.shipping;

import java.util.ArrayList;
import java.util.HashSet;

import javax.inject.Inject;

import nl.omniex.omniexshopping.data.model.address.Address;
import nl.omniex.omniexshopping.data.repository.AddressRepository;
import nl.omniex.omniexshopping.ui.base.BasePresenter;
import timber.log.Timber;

public class OrderShippingPresenter extends BasePresenter<OrderShippingView> {

    private AddressRepository mAddressRepository;

    @Inject
    OrderShippingPresenter(AddressRepository addressRepository) {
        mAddressRepository = addressRepository;
    }

    void getShippingAddresses() {
        addDisposable(mAddressRepository
                .getShippingAddresses()
                .subscribe(
                        shippingAddressesResponse -> {
                            ifViewAttached(view -> view.onShippingAddressesFetched(new ArrayList<>(new HashSet<>(shippingAddressesResponse.body().getShippingAddressList().getAddressList()))));
                        }, error->{
                            ifViewAttached(OrderShippingView::onGetShippingAddressesFailed);
                            Timber.e(error);
                        }));
    }

    void setShippingAddress(Address existingAddress) {
        addDisposable(mAddressRepository
                .setExistingShippingAddress(existingAddress)
                .subscribe(voidResponse -> {
                    ifViewAttached(view -> view.onShippingAddressSet());
                }, Throwable::printStackTrace));
    }
}

File: app/src/main/java/omniex/nl/omniex/ui/app/order/shipping/OrderShippingPresenter.kt
package omniex.nl.omniex.ui.app.order.shipping

import omniex.nl.omniex.data.model.address.Address
import omniex.nl.omniex.data.repository.AddressRepository
import omniex.nl.omniex.ui.base.BasePresenter
import timber.log.Timber
import javax.inject.Inject

class OrderShippingPresenter @Inject
internal constructor(private val mAddressRepository: AddressRepository) : BasePresenter<OrderShippingView>() {

    internal fun getShippingAddresses() {
        addDisposable(mAddressRepository
                .shippingAddresses()
                .subscribe(
                        { shippingAddressesResponse -> ifViewAttached { view -> view.onShippingAddressesFetched(shippingAddressesResponse.body()!!.shippingAddressList!!.addressList!!) } },
                        { error ->
                            ifViewAttached({ it.onGetShippingAddressesFailed() })
                            Timber.e(error)
                        }))
    }

    internal fun setShippingAddress(existingAddress: Address) {
        addDisposable(mAddressRepository
                .setExistingShippingAddress(existingAddress)
                .subscribe({ voidResponse -> ifViewAttached { view -> view.onShippingAddressSet() } }, {it.printStackTrace()}))
    }
}
